### PR TITLE
fix: retry gRPC operations if the connection is unexpectedly terminated

### PR DESCRIPTION
**Pull request by mr-salty**
part of #4714

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087)

### CODE REVIEW

**Review by coryan — APPROVED** [review]
This is fine too. Up to you.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-:-MHXm2FxB2s7pZiGaXEp:b-riykxk)* status: :shipit: complete! all files reviewed, all discussions resolved

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 1]
```diff
@@ -1469,6 +1469,43 @@ TEST(ConnectionImplTest, ExecutePartitionedDmlDeleteTooManyTransientFailures) {
...
+
+  auto constexpr kTextTxn = R"pb(
+    id: "1234567890"
+  )pb";
+  spanner_proto::Transaction txn;
+  ASSERT_TRUE(TextFormat::ParseFromString(kTextTxn, &txn));
+  EXPECT_CALL(*mock, BeginTransaction(_, _)).WillOnce(Return(txn));
+
+  auto constexpr kTextResponse = R"pb(
+    metadata: { transaction: { id: "2345678901" } }
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc:1488]: Q: Why are there two transaction IDs?

File: google/cloud/spanner/retry_policy.h [hunk 2]
```diff
@@ -32,8 +33,23 @@ struct SafeGrpcRetry {
...
-           status.code() == StatusCode::kResourceExhausted;
+    if (status.code() == StatusCode::kUnavailable ||
+        status.code() == StatusCode::kResourceExhausted) {
+      return true;
+    }
+    // Treat the unexpected termination of the gRPC connection as retryable.
+    // There is no explicit indication of this, but it will result in an
+    // INTERNAL status with one of the `kRetryableMessages`.
+    if (status.code() == StatusCode::kInternal) {
+      constexpr char const* kRetryableMessages[] = {
```
Comment (devbww) [inline google/cloud/spanner/retry_policy.h:44]: Nit: Given that this is `IsTransientFailure()`, I might choose to call these `kTransientMessages`?

**Review by mr-salty — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087)* status: all files reviewed, 2 unresolved discussions (waiting on @devbww)

___
*[google/cloud/spanner/retry_policy.h, line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0b-r1-44:-MHY6AL-85TxNcOu1OOU:b-n9rhvb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/retry_policy.h#L44)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: Given that this is `IsTransientFailure()`, I might choose to call these `kTransientMessages`?
</blockquote></details>

Done. I thought `kTransientFailureMessages` read a little better and more importantly did not cause any extra line wraps!
___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 1432 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0a-r1-1432:-MHY6KyF0X1yjaGnrUTM:bmf11j7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/internal/connection_impl_test.cc#L1432)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

s/not a retryable/a non-retryable/ perhaps?
</blockquote></details>

This is my usual riff on "ceci n'est pas une pipe"... but I did change it to "transient".
___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 1488 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0a-r1-1488:-MHY75XMDreyLL2tx3Xs:bkxf0bb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/internal/connection_impl_test.cc#L1488)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Q: Why are there two transaction IDs?
</blockquote></details>

hm, I missed changing one... but, that second one (in the response) shouldn't be there at all, since PDML always explicitly does `BeginTransaction`.  It will just get ignored since we don't even look at it in `ConnectionImpl`. So, I deleted it here and in the place I copied it from.

**Review by devbww — COMMENTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087)* status: all files reviewed, 1 unresolved discussion (waiting on @mr-salty)

___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 1432 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0a-r1-1432:-MHYSDRR4zY8WygXJC1G:biw8avb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/internal/connection_impl_test.cc#L1432)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

This is my usual riff on "ceci n'est pas une pipe"... but I did change it to "transient".
</blockquote></details>

My comment was more about clarity.  It certainly is an INTERNAL error, so saying it is not one sounds strange.  That is, I wanted to bound the "not" closely to the "transient", and use a positive assertion ... "this is a non-transient INTERNAL error".

**Review by mr-salty — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 1432 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0a-r1-1432:-MHlxAidBZ9T92A1m3ms:bfbsqut) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/internal/connection_impl_test.cc#L1432)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

My comment was more about clarity.  It certainly is an INTERNAL error, so saying it is not one sounds strange.  That is, I wanted to bound the "not" closely to the "transient", and use a positive assertion ... "this is a non-transient INTERNAL error".
</blockquote></details>

I think you're reading too much into a mostly opaque string, but the underlying issue subtle enough to warrant a comment or two, which i added.

**Review by devbww — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087)* status: 1 of 2 files reviewed, all discussions resolved (waiting on @devbww)

___
*[google/cloud/spanner/internal/connection_impl_test.cc, line 1432 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5087#-MHXdVfB9cQHrmBX9N0a-r1-1432:-MHlzi5s5Xwp4wJDYVGg:b-qbe78d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4e806b4fc955bdd9202ad3f41c7722f2060f8bc7/google/cloud/spanner/internal/connection_impl_test.cc#L1432)):*
<details><summary><i>Previously, mr-salty (Todd Derr) wrote…</i></summary><blockquote>

I think you're reading too much into a mostly opaque string, but the underlying issue subtle enough to warrant a comment or two, which i added.
</blockquote></details>

OK, "non-transient" -> "permanent" works for me.

**Pull request merged**